Sources/Danger/GitDiff.swift

```diff
@@ -0,0 +1,169 @@
+            }
+        }
+    }
+}
+
+extension FileDiff.Line {
+    enum ChangeType: Equatable {
+        case added
+        case removed
+        case unchanged
+    }
+}
+
+extension StringProtocol {
+    func deletingPrefix(_ prefix: String) -> String {
+        guard hasPrefix(prefix) else { return String(self) }
+        return String(dropFirst(prefix.count))
+    }
+}
+
+struct DiffParser {
+    func parse(_ diff: String) -> [FileDiff] {
+        diff.components(separatedBy: "diff --git ").compactMap { fileDiff in
+            let headerAndHunks = fileDiff.components(separatedBy: "@@")
+            guard let header = headerAndHunks.first,
+                !header.isEmpty else {
+                return nil
+            }
+
+            return FileDiff(parsedHeader: parseHeader(header), hunks: parseHunks(Array(headerAndHunks.dropFirst())))
+        }
+    }
+
+    private func parseHeader(_ header: String) -> FileDiff.ParsedHeader {
+        let lines = header.components(separatedBy: "\n")
+
+        let filePath = lines.first?.split(separator: " ").first(where: { $0.starts(with: "b/") })?.deletingPrefix("b/") ?? ""
+
+        let change: FileDiff.ParsedHeader.ChangeType
+
+        if lines.contains(where: { $0.hasPrefix("deleted file mode ") }) {
+            change = .deleted
+        } else if lines.contains(where: { $0.hasPrefix("new file mode") }) {
+            change = .created
+        } else if let modifiedLineIndex = lines.firstIndex(where: { $0.hasPrefix("rename from ") }) {
+            change = .renamed(oldPath: lines[modifiedLineIndex].deletingPrefix("rename from "))
+        } else {
+            change = .modified
+        }
+
+        return FileDiff.ParsedHeader(filePath: filePath, change: change)
+    }
+
+    private func parseHunks(_ hunks: [String]) -> [FileDiff.Hunk] {
+        (0 ..< hunks.count / 2).compactMap { index -> FileDiff.Hunk? in
+            let changesSpan = hunks[index * 2]
+            let changes = hunks[index * 2 + 1]
+            let lines = changes.components(separatedBy: "\n").dropFirst().dropLast().map(parseLine)
+            let parsedChanges = parseChangesSpan(changesSpan)
+
+            return FileDiff.Hunk(oldLineStart: parsedChanges.oldLineStart, oldLineSpan: parsedChanges.oldLineSpan, newLineStart: parsedChanges.newLineStart, newLineSpan: parsedChanges.newLineSpan, lines: lines)
+        }
+    }
+
+    private func parseLine(_ line: String) -> FileDiff.Line {
+        let text = String(line.dropFirst())
+        if line.hasPrefix("+") {
+            return FileDiff.Line(text: String(line.dropFirst()), changeType: .added)
+        } else if line.hasPrefix("-") {
+            return FileDiff.Line(text: String(line.dropFirst()), changeType: .removed)
+        } else {
+            return FileDiff.Line(text: text, changeType: .unchanged)
+        }
+    }
+
+    private func parseChangesSpan(_ changesSpan: String) -> (oldLineStart: Int, oldLineSpan: Int, newLineStart: Int, newLineSpan: Int) {
```

Comment:
<!--
  1 failure:  Tuples should hav...
  1 warning:  Line should be 12...
  
  
  DangerID: danger-id-Danger;
  File: Sources/Danger/GitDiff.swift;
  Line: 155;
-->
- :no_entry_sign: Tuples should have at most 2 members. (`large_tuple`)
- :warning: Line should be 120 characters or less: currently 136 characters (`line_length`)